### PR TITLE
[SILOpt] Hoist destroys for non-lexical alloc_stacks more aggressively.

### PR DESCRIPTION
**Pull request by nate-chandler**
Previously, when hoisting `destroy_addr`s for `alloc_stack`s, deinit barriers were always considered barriers to hoisting.  Here, that is changed so that hoisting only is barred when the `alloc_stack` whose `destroy_addr` is being hoisted is lexical.

Additionally, more non-lexical `alloc_stack`s are formed: only `alloc_box`es whose lifetimes are protected via a lexical borrow scope get promoted to lexical `alloc_stack`s.  This will become more significant when types and values are annotated `@_eagerMove`.

In order to do this safely, we must ensure that implicit lexical lifetimes are not removed.  Address function arguments with a convention other than Indirect_Inout have implicit lexical lifetimes: e.g. SSADestroyHoisting respects deinit barriers when hoisting the the destroy_addrs of such arguments.  Until we have `begin_borrow_addr`/`end_borrow_addr`, when inlining functions with such arguments, if the address passed to it has an `alloc_stack` as its storage base, mark that `alloc_stack` lexical.

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please smoke test macOS platform

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

File: lib/SILOptimizer/Transforms/AllocBoxToStack.cpp [hunk 1]
```diff
@@ -538,13 +540,31 @@ static bool rewriteAllocBoxAsAllocStack(AllocBoxInst *ABI) {
...
+    bool lexicalLifetimesEnabled =
+        mod.getASTContext().SILOpts.supportsLexicalLifetimes(mod);
+    if (!lexicalLifetimesEnabled)
+      return false;
+    // Look for lexical borrows of the alloc_box.
+    GraphNodeWorklist<Operand *, 4> worklist;
+    worklist.initializeRange(ABI->getUses());
+    while (auto *use = worklist.pop()) {
+      // See through mark_uninitialized instructions.
+      if (auto *mui = dyn_cast<MarkUninitializedInst>(use->getUser())) {
```
Comment (atrick) [inline lib/SILOptimizer/Transforms/AllocBoxToStack.cpp:556]: Hard-coding opcodes here (for mark_uninitialized) creates an implicit correctness dependency on SIL design. How will we find out if SILGen ever introduces new instructions in the same position as mark_uninitialized? Maybe we want a flag on alloc_box just at least as a verification mechanism?

**Review by atrick — APPROVED** [review]
Looks great! Thanks.

File: lib/SILOptimizer/Transforms/AllocBoxToStack.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/SILOptimizer/Transforms/AllocBoxToStack.cpp:556]: Added verification that `begin_borrow [lexical]`s whose operand's type is a `SILBoxType` is either an `AllocBoxInst` or a `SILFunctionArgument` (looking through `copy_value`s, `begin_borrow`s, and `mark_uninitialized` instructions) in https://github.com/apple/swift/pull/60478/commits/fdb20e1b8dad188861cfe4ebeb038f4145952fe6 .

Comment (nate-chandler) [reply lib/SILOptimizer/Transforms/AllocBoxToStack.cpp:558]: AllocBoxToStack also now looks through to the uses of non-lexical begin_borrows to properly handle the case of repeated inlining where the first callee's SILFunctionArgument is non-lexical but the second callee's SILFunctionArgument is lexical in https://github.com/apple/swift/pull/60478/commits/896a464ced0a7baaf30a202a58ed060ecfde4efa .

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Pull request merged**